### ISSUE TITLE
Performance regression on some models

### ISSUE THREAD

**Issue by peastman** [openmm/openmm#4334]
In #4147 we added an optimization to reduce the O(n^2) cost of building the neighbor list for very large systems.  Unfortunately, it was incompatible with another optimization we were already doing: sorting atom blocks by size to produce a more compact neighbor list.  We have to choose between one optimization and the other, one reducing the cost of building the neighbor list and one reducing the cost of computing interactions.  Since one scales as O(n^2) and the other as O(n), which one is more important depends on the size of the system.  We ran a bunch of tests and selected 90,000 atoms as the cutoff for selecting between the two.

It turns out this is very sensitive to the details of the system.  Even for some systems much larger than that, you're much better off sorting the blocks rather than trying to optimize building the neighbor list.  A common situation is when you have a mix of water and ions.  The blocks containing the ions are much larger than the ones containing the water (many fewer atoms scattered over the same volume), so it's very important for them to be sorted to the end.  When OpenMM builds water boxes it puts the ions at the end, right where we want them, so additional sorting isn't needed.  But when Amber builds water boxes, it puts the ions before the water, not after.

To show the impact of this, I ran benchmarks of water boxes of varying sizes using the script from https://github.com/openmm/openmm/issues/4063#issuecomment-1536768880, only specifying `ionicStrength=0.1*molar`.  I then did the same, only reversing the order of the water and ions:

```python
modeller = Modeller(Topology(), [])
modeller.addSolvent(ff, boxSize=Vec3(width, width, width)*nanometers, ionicStrength=0.1*molar)
m2 = Modeller(modeller.topology, modeller.positions)
m2.delete([r for r in m2.topology.residues() if r.name != 'HOH'])
modeller.deleteWater()
modeller.add(m2.topology, m2.positions)
```

Here is the time per 100,000 atoms to run 5000 steps.  Lower is better.

|Box Width|Atoms|w/o large blocks|Large blocks|Large blocks, salt reversed|
|---|---|---|---|---|
5|12227|7.44|7.35|12.15
10|98644|2.63|2.61|4.68
15|334217|2.67|2.60|3.31
20|790542|3.06|2.70|2.95
25|1546428|4.57|3.79|3.56

The results for the standard water box are very similar to what we found in #4147: the large blocks optimization improves performance in all cases, with the amount of improvement growing as the system grows.  But reversing the order to put the ions before the water dramatically changes it.  The smaller systems get much slower.  Even at 334,000 atoms, using the large blocks optimization slows things down significantly.  You have to get up to around 700,000 atoms before it becomes faster than omitting the large blocks optimization and instead sorting blocks by size.

**Comment by ex-rzr** [comment]
Is "w/o large blocks" the same for both original and reversed orders?

**Comment by peastman** [comment]
Right.  In that case it sorts the atom blocks by size, so the original order doesn't matter.

I have an idea for a fix I'm working on.  I hope to get a draft PR in soon.

**Comment by peastman** [comment]
The fix is in #4343.
